src/Shared/IInternable.cs

```diff
@@ -0,0 +1,346 @@
+            // For example, C:\project1, C:\project2
+            for (int i = other.Length - 1; i >= 0; --i)
+            {
+                if (_target[i + _startIndex] != other[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Don't use this function. Use ExpensiveConvertToString
+        /// </summary>
+        public override string ToString()
+        {
+            throw new InvalidOperationException();
+        }
+    }
+
+    /// <summary>
+    /// Wrapper over a string.
+    /// </summary>
+    internal struct StringInternTarget : IInternable
+    {
+        /// <summary>
+        /// Stores the wrapped string.
+        /// </summary>
+        private readonly string _target;
+
+        /// <summary>
+        /// Constructor of the class
+        /// </summary>
+        /// <param name="target">The string to wrap</param>
+        internal StringInternTarget(string target)
+        {
+            ErrorUtilities.VerifyThrowArgumentLength(target, nameof(target));
+            _target = target;
+        }
+
+        /// <summary>
+        /// Gets the length of the target string.
+        /// </summary>
+        public int Length => _target.Length;
+
+        /// <summary>
+        /// Gets the n character in the target string.
+        /// </summary>
+        /// <param name="index">Index of the character to gather.</param>
+        /// <returns>The character in the position marked by index.</returns>
+        public char this[int index] => _target[index];
+
+        /// <summary>
+        /// Returns the target which is already a string.
+        /// </summary>
+        /// <returns>The target string.</returns>
+        public string ExpensiveConvertToString() => _target;
+
+        /// <summary>
+        /// Compare target to string. Assumes string is of equal or smaller length than target.
+        /// </summary>
+        /// <param name="other">The string to compare with the target.</param>
+        /// <returns>True if target starts with <paramref name="other"/>, false otherwise.</returns>
+        public bool StartsWithStringByOrdinalComparison(string other) => _target.StartsWith(other, StringComparison.Ordinal);
+
+        /// <summary>
+        /// Verifies if the reference of the target string is the same of the given string.
+        /// </summary>
+        /// <param name="other">The string reference to compare to.</param>
+        /// <returns>True if both references are equal, false otherwise.</returns>
+        public bool ReferenceEquals(string other) => ReferenceEquals(_target, other);
+    }
+
+    /// <summary>
+    /// Wrapper over a substring of a string.
+    /// </summary>
+    internal struct SubstringInternTarget : IInternable
+    {
+        /// <summary>
+        /// Stores the wrapped string.
```

Comment:
This seems like maybe it should be rewritten in terms of `ReadOnlySpan<char>`'s. I see that MSBuild already has taken a dependency on System.Memory so it is available.

Reply:
Will be addressed in a separate PR.